File: android/src/main/kotlin/io/rodcone/flutter_meta_wearables_dat/StreamSessionStateStreamHandler.kt
package io.rodcone.flutter_meta_wearables_dat

import com.meta.wearable.dat.camera.StreamSession
import com.meta.wearable.dat.camera.types.StreamSessionState
import io.flutter.plugin.common.EventChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch

/**
 * Stream handler for stream session state updates from the DAT SDK.
 * Emits integer values matching the Dart `StreamSessionState` enum.
 *
 * Set the [session] property when a stream session is created.
 * Clear it when the session is torn down.
 */
internal class StreamSessionStateStreamHandler : EventChannel.StreamHandler {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private var job: Job? = null
    private var eventSink: EventChannel.EventSink? = null

    /** The active stream session to observe. Setting this resubscribes to the state flow. */
    var session: StreamSession? = null
        set(value) {
            field = value
            resubscribe()
        }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        if (events == null) return
        eventSink = events
        resubscribe()
    }

    override fun onCancel(arguments: Any?) {
        job?.cancel()
        job = null
        eventSink = null
    }

    private fun resubscribe() {
        job?.cancel()
        job = null
        val sink = eventSink ?: return
        val currentSession = session ?: return

        job =
                scope.launch {
                    currentSession.state.collect { state -> sink.success(mapState(state)) }
                }
    }

    fun dispose() {
        job?.cancel()
        job = null
        eventSink = null
        scope.cancel()
    }

    companion object {
        /**
         * Maps Android SDK StreamSessionState to int values expected by Dart.
         *
         * Android enum: STARTING, STARTED, STREAMING, STOPPING, STOPPED, CLOSED
         * Dart enum:    stopping(0), stopped(1), waitingForDevice(2), starting(3), streaming(4), paused(5)
         *
         * Mapping notes:
         * - STARTED has no direct Dart equivalent; mapped to starting(3) since it precedes streaming.
         * - CLOSED has no direct Dart equivalent; mapped to stopped(1).
         * - Android has no WAITING_FOR_DEVICE or PAUSED states.
         */
        fun mapState(state: StreamSessionState): Int {
            return when (state) {
                StreamSessionState.STOPPING -> 0
                StreamSessionState.STOPPED -> 1
                StreamSessionState.STARTING -> 3
                StreamSessionState.STARTED -> 3  // No Dart equivalent; treat as "starting"
                StreamSessionState.STREAMING -> 4
                StreamSessionState.CLOSED -> 1   // No Dart equivalent; treat as "stopped"
            }
        }
    }
}

File: ios/Classes/StreamSessionStateStreamHandler.swift
import Flutter
import MWDATCamera
import MWDATCore

/// Stream handler for stream session state updates.
///
/// Set the `session` property when a stream session is created.
/// Clear it when the session is torn down.
class StreamSessionStateStreamHandler: NSObject, FlutterStreamHandler {
  /// The active stream session to observe. Setting this property
  /// re-subscribes to the session's state publisher.
  var session: StreamSession? {
    didSet { resubscribe() }
  }

  private var eventSink: FlutterEventSink?
  private var listenerToken: AnyListenerToken?

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    resubscribe()
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    cancelListener()
    return nil
  }

  private func resubscribe() {
    cancelListener()
    guard let session = session, let events = eventSink else { return }

    // Send current state immediately
    events(Self.stateToInt(session.state))

    // Listen for future changes
    listenerToken = session.statePublisher.listen { state in
      Task { @MainActor in
        events(Self.stateToInt(state))
      }
    }
  }

  private func cancelListener() {
    if let token = listenerToken {
      Task { await token.cancel() }
      listenerToken = nil
    }
  }

  /// Maps StreamSessionState to the int values expected by Dart.
  private static func stateToInt(_ state: StreamSessionState) -> Int {
    switch state {
    case .stopping:         return 0
    case .stopped:          return 1
    case .waitingForDevice: return 2
    case .starting:         return 3
    case .streaming:        return 4
    case .paused:           return 5
    @unknown default:       return 1 // default to stopped
    }
  }
}
